### ISSUE TITLE
ConditionalMNLogit fit() Method Still Executing After 9+ Hours

### ISSUE THREAD

**Issue by dave660** [statsmodels/statsmodels#7359]
Was attempting to use the ConditionalMNLogit classifier on a dataset to has a dependent variable that can vary from 1-12, but not all groups will have 1-12 and instead can have any number within that range. The problem I am seeing is that the call to the fit() method executes and never completes....at least after running up to 9 hours. Below is the code being used. Any insight would be greatly appreciated!

from sklearn import preprocessing
from numpy import genfromtxt, savetxt
import numpy as np
import datetime
import pandas as pd
from sklearn.metrics import accuracy_score
from hdcpping_imports import get_dropped_cols_list
from statsmodels.discrete import conditional_models as cm
from sklearn.preprocessing import StandardScaler

inputFile = "./Data/BrisModelRF.csv"
testInputFile = "./Data/BrisModelRF_Test.csv"
linerRegressionModelFileName = "LR_compressed.joblib"
dropped_cols_file = "./Data/DropColumns.csv"

df = pd.read_csv(inputFile,sep=',',header=0)
df['RaceDate'] = pd.to_datetime(df['RaceDate'])
df_test = pd.read_csv(testInputFile,sep=',',header=0)
df_test['RaceDate'] = pd.to_datetime(df_test['RaceDate'])

dropped_cols_list = get_dropped_cols_list(dropped_cols_file)

y = df['FinishPosition'] -1
X = df.drop(dropped_cols_list, axis=1)
#print(X.head)
mnl = cm.ConditionalMNLogit(y, X, groups = df["RaceId"], full_output = True)
mnl_result = mnl.fit()
print(mnl_result.summary())

**Comment by kshedden** [comment]
Conditional inference has exponential complexity in the group sizes.  So it is only practically usefull if the groups are not too large.

**Comment by dave660** [comment]
> Conditional inference has exponential complexity in the group sizes. So it is only practically usefull if the groups are not too large.

So 12 groups would be considered too large?

**Comment by kshedden** [comment]
It's the group sizes, not the number of groups.  What is the size of your largest group?

Groups here refers to the variable that you are clustering on, RaceId, not the number of levels in the multinomial outcome.

**Comment by dave660** [comment]
> It's the group sizes, not the number of groups. What is the size of your largest group?
> 
> Groups here refers to the variable that you are clustering on, RaceId, not the number of levels in the multinomial outcome.

Ok got  it....there are only 85 different groups in this dataset.

**Comment by kshedden** [comment]
But it's not the number of groups that matters, it is the number of observations within a group.  What is the largest group?

**Comment by dave660** [comment]
> But it's not the number of groups that matters, it is the number of observations within a group. What is the largest group?

In this case up to 12 in a group...most are smaller and probably 7-8 on average.

**Comment by kshedden** [comment]
Here are two more suggestions: First, make sure that all of your covariates vary within groups.  If you have any covariates that are constant within all groups the model is not identified.  Second, run it again with maxiter=1 to see if the issue is time-per-iteration or the number of iterations.

**Comment by dave660** [comment]
> Here are two more suggestions: First, make sure that all of your covariates vary within groups. If you have any covariates that are constant within all groups the model is not identified. Second, run it again with maxiter=1 to see if the issue is time-per-iteration or the number of iterations.

OK I will try your suggestion. Thanks for your help and I will let you know what happens when setting maxiter=1.

**Comment by dave660** [comment]
> > Here are two more suggestions: First, make sure that all of your covariates vary within groups. If you have any covariates that are constant within all groups the model is not identified. Second, run it again with maxiter=1 to see if the issue is time-per-iteration or the number of iterations.
> 
> OK I will try your suggestion. Thanks for your help and I will let you know what happens when setting maxiter=1.

Looks like even with the maxiter=1 specified this runs for an extended amount of time without ever completing the fit() method call. Looks like it is on to R to see if I can get this to work in that language. Was really hoping to stay with Python though as I am familiar with this language already. Thanks again for your help with this anyway.

**Comment by kshedden** [comment]
The conditional log likelihood involves a sum over m! terms, where m is the group size.  With m=12, that is almost half a billion terms, just within one group for one likelihood evaluation.

**Comment by dave660** [comment]
Wow...not surprising then that is a very long execution time.  Thanks again for your help!
